datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -311,20 +311,27 @@ fn stats_cartesian_product(
     }
 }
 
-/// A stream that issues [RecordBatch]es as they arrive from the right  of the join.
+/// A stream that issues [RecordBatch]es as they arrive from the right of the join.
+/// Right column orders are preserved.
 struct CrossJoinStream {
     /// Input schema
     schema: Arc<Schema>,
-    /// future for data from left side
+    /// Future for data from left side
     left_fut: OnceFut<JoinLeftData>,
-    /// right
+    /// Right stream
     right: SendableRecordBatchStream,
-    /// Current value on the left
-    left_index: usize,
-    /// Current batch being processed from the right side
-    right_batch: Arc<parking_lot::Mutex<Option<RecordBatch>>>,
-    /// join execution metrics
+    /// Join execution metrics
     join_metrics: BuildProbeJoinMetrics,
+    /// State information
+    state: CrossJoinStreamState,
+    /// Left data
+    left_data: Vec<RecordBatch>,
+    /// Current right batch
+    right_batch: RecordBatch,
+    /// Indexes the next processed build side batch
+    left_batch_index: usize,
```

Comment:
minor: Could batch & row indices be a contents of GenerateResult state (as they seem to be valid only while result generation)?

Reply:
`build_batch` also uses them. I have tried it now but the current version seems more simple rather than adding them to function parameters as `&mut` or unwrapping the state to extract them.